### PR TITLE
Fix memory leak

### PR DESCRIPTION
**Pull request by dead-horse**
## Description

**What changed?**

- Clean timeout tasks when process wait task queue.
- End session when query error.

Try to fix https://jira.mongodb.org/browse/NODE-2413?focusedCommentId=3094451&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-3094451

### CODE REVIEW

File: lib/cursor/core_cursor.js [hunk 1]
```diff
@@ -434,6 +434,10 @@ class CoreCursor extends Readable {
     }
 
     function done(err, result) {
+      if (err && cursor._endSession) {
```
Comment (dead-horse) [inline lib/cursor/core_cursor.js:437]: After I monkey patched mongodb with `processWaitQueue` and deployed in my environment, when this happen again, the queue is no more memory leak if I set `waitQueueTimeoutMS`. But there's another place where there's a memory leak.

Then I take a heapdump

image
image

But I'm not sure if we can fix this memory leak like this.

**Comment by dead-horse** [comment]
@mbroadst any thoughts?

**Comment by mbroadst** [comment]
@dead-horse thank you very much for the fix here, I think you've indeed found some bugs. I will be working on this later today, and try to include some test cases. The fix will go out in next week's patch release.

**Comment by dead-horse** [comment]
@mbroadst  Any progress? Do we need to provide more information?

**Comment by mbroadst** [comment]
@dead-horse the fix will be merged in as part of #2388 , and put into a patch release today. Thank you for your patience.

**Comment by dead-horse** [comment]
thanks @mbroadst 

#2388 fixed the wait queue problem, but I'm not sure how the cursor problem arose and how to solve it.

**Comment by mbroadst** [comment]
@dead-horse 3.5.8 has been released with a fix for the memory leak in the wait queue (#2388) and the implicit session issue (#2392). I'm going to close this PR now, let's speak on JIRA if you continue to experience issues.

**Pull request closed without merging**